src/dmd/cparse.d

```diff
@@ -5090,4 +5112,63 @@ final class CParser(AST) : Parser!AST
     }
 
     //}
+
+    /******************************************************************************/
+    /********************************* #define Parser *****************************/
+    //{
+
+    /**
+     * Go through the #define's in the defines buffer and see what we can convert
+     * to Dsymbols, which are then appended to symbols[]
+     */
+    void addDefines()
```

Comment:
This should really be a different file. It's going to get very complicated and should be using the C parser versus being part of the C parser.

Reply:
I've toyed with the idea of making it a class derived from `CParser`. But it's a lot easier to make architectural decisions once we know what we're doing and how it's going to interact with `CParser`. So in the meantime I went for the most direct, straightforward approach.

Reply:
I still had to re-architect that section several times, even though it doesn't do much.

Reply:
It shouldn't be related by blood to CParser but rather be dependant on CParser, for manifest constants at least.

Reply:
We won't really know what it needs from CParser until it is better developed. For example, some of the pattern matching can make use of "is the expansion a type", "is the expansion an expression", etc. It's much more than manifest constants.

Reply:
After all, the macro expansions are fragments of C syntax.

Reply:
We already have mixin types and expressions in D, what won't work by just reusing them to represent macros?

Reply:
> We already have mixin types and expressions in D, what won't work by just reusing them to represent macros? 

Yes, I'm well aware of the possibility to be tricksy with macros. :-)
```
// #define TYPE fun(); int
TYPE var;
mixin("fun(); int") var;
```

Reply:
```
#define TYPE int long
```
How is that going to work with mixins? (I see similar macros cpp predefines.) I.e. how are you going to detect it is a type so you can generate the right mixin for it? I plan on using `CParser.isTypeName()` and those other `is` functions in CParser to detect what kind of thing the replacement text is.

Reply:
This one should be trivially turned into a symbol directly but a mixin can be used in place of a type.

Also "generate the right mixin" sounds like a bad mixture of the two solutions. Mixin is a fallback, generation of symbols is supposed to be enough.

Reply:
Yeah, but you have to know it is a type in order to generate the right mixin. As for generating symbols, that's exactly what this PR does.

Reply:
We don't have to generate the right mixin as per se.

The mixins for types, decls, and expressions are done in different AST nodes, no?